### PR TITLE
Add `linalg.qr` to ChainerX

### PR DESCRIPTION
**Pull request by IvanYashchuk**
This is work-in-progress pull request!

This PR adds QR decomposition providing bindings to [cuSOLVER's QR decomposition](https://docs.nvidia.com/cuda/cusolver/index.html#cuds-lt-t-gt-geqrf) on GPU and to LAPACK on CPU.

With reference to #6423  and #6764

### CODE REVIEW

**Comment by IvanYashchuk** [comment]
QR decomposition is working now correctly.

File: chainerx_cc/chainerx/cuda/cuda_device/linalg.cu [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+        Device& device = a.device();
+        Dtype dtype = a.dtype();
+        CudaSetDeviceScope scope{device.index()};
+
+        if (a.ndim() != 2) {
+            throw DimensionError{"ChainerX QR supports only 2-dimensional arrays."};
+        }
+
+        int m = a.shape()[0];
+        int n = a.shape()[1];
```
Comment (hvy) [inline chainerx_cc/chainerx/cuda/cuda_device/linalg.cu]: Let's be explicit with `int64_t`.

**Comment by IvanYashchuk** [comment]
Currently blocked by #7791

**Comment by hvy** [comment]
The `tri*` routines have now been merged. Could you rebase/update this PR?

File: chainerx_cc/chainerx/routines/linalg.h [hunk 2]
```diff
@@ -13,4 +15,14 @@ Array Solve(const Array& a, const Array& b);
 
 Array Inverse(const Array& a);
 
+enum class QrMode {
+    // if K = min(M, N), where `a` of shape (M, N)
+    reduced,  // returns q, r with dimensions (M, K), (K, N) (default)
+    complete,  // returns q, r with dimensions (M, M), (M, N)
+    r,  // returns empty q and r with dimensions (0, 0), (K, N)
+    raw  // returns h, tau with dimensions (N, M), (K, 1)
```
Comment (hvy) [inline chainerx_cc/chainerx/routines/linalg.h]: Could you follow the existing enum convention which is to have values named `kReduced`, `kComplete`, etc. ("k" for "constant")?

File: chainerx_cc/chainerx/routines/linalg.cc [hunk 3]
```diff
@@ -189,4 +190,79 @@ Array Inverse(const Array& a) {
...
+        q_shape = Shape{m, k};
+        r_shape = Shape{k, n};
+    } else if (mode == QrMode::complete) {
+        q_shape = Shape{m, m};
+        r_shape = Shape{m, n};
+    } else if (mode == QrMode::r) {
+        r_shape = Shape{k, n};
+    } else if (mode == QrMode::raw) {
+        r_shape = Shape{n, m};  // for "raw" mode r in the code corrensponds to h in docs
+    }
```
Comment (hvy) [inline chainerx_cc/chainerx/routines/linalg.cc]: How about using switch cases here and in the Python bindings?

File: chainerx_cc/chainerx/routines/linalg.cc [hunk 4]
```diff
@@ -189,4 +190,87 @@ Array Inverse(const Array& a) {
...
+            break;
+        case QrMode::kR:
+            r_shape = Shape{k, n};
+            break;
+        case QrMode::kRaw:
+            r_shape = Shape{n, m};  // for "raw" mode r in the code corrensponds to h in docs
+            break;
+        default:
+            throw ChainerxError{"Invalid QR mode"};
+            break;
```
Comment (IvanYashchuk) [inline chainerx_cc/chainerx/routines/linalg.cc]: Without `break` statements I get compiler errors
```
error: this statement may fall through [-Werror=implicit-fallthrough=]
```

File: chainerx_cc/chainerx/routines/linalg.cc (hunk 3 above)
Comment (IvanYashchuk) [reply chainerx_cc/chainerx/routines/linalg.cc]: Done here.
In Python bindings the conditional statements are with `mode` of type `std::string`, switch cases are not applicable there, right?

File: chainerx_cc/chainerx/routines/linalg.cc [hunk 5]
```diff
@@ -188,4 +189,87 @@ Array Inverse(const Array& a) {
...
+                const Array& a = bctx.GetRetainedInput(a_tok);
+                const Array& Q = bctx.GetRetainedOutput(q_tok);
+                const Array& R = bctx.GetRetainedOutput(r_tok);
+
+                if (R.shape()[0] != R.shape()[1]) {
+                    throw DimensionError{"ChainerX QR differentiation is not implemented for non-square R."};
+                }
+
+                const Array& dQ = bctx.output_grad(0).has_value() ? *bctx.output_grad(0) : Zeros(a.shape(), a.dtype(), a.device());
+                const Array& dR = bctx.output_grad(1).has_value() ? *bctx.output_grad(1) : Zeros(a.shape(), a.dtype(), a.device());
```
[the remark this answers is not in the document]
Comment (IvanYashchuk) [reply chainerx_cc/chainerx/routines/linalg.cc]: Fixed that.

File: tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py [hunk 6]
```diff
@@ -231,3 +231,73 @@ def forward_xp(self, inputs, xp):
...
+    chainer.testing.product({
+        'shape': [(1, 1), (6, 6)],
+        'in_dtypes': ['float32', 'float64'],
+        'mode': ['reduced', 'complete']
+    }) + chainer.testing.product({
+        'shape': [(3, 2)],
+        'in_dtypes': ['float32', 'float64'],
+        'mode': ['reduced']
+    })
+))
```
[the remark this answers is not in the document]
Comment (IvanYashchuk) [reply tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py]: Input of shape `(0, N)` now works for all modes.
There is a problem of different strides for `r` matrix (compared to NumPy) for shape of `(N, 0)` .
Also with shape `(N, 0)` and `mode='complete'` cuSOLVER does not return the same result as LAPACK for `q` matrix (it should be the identity matrix).

[the remark this answers is not in the document]
Comment (IvanYashchuk) [reply tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py]: New tests fail in Travis. The reason is that older NumPy versions did not allow empty arrays.
```
=============================================== short test summary info ================================================
FAIL tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py::test_QR_param_26_{shape=(0, 3), mode='complete', skip_backward_test=True, skip_double_backward_test=True, in_dtypes='float64'}[native:0]
FAIL tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py::test_QR_param_10_{shape=(0, 3), mode='r', skip_backward_test=True, skip_double_backward_test=True, in_dtypes='float64'}[native:0]
FAIL tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py::test_QR_param_5_{shape=(0, 3), mode='raw', skip_backward_test=True, skip_double_backward_test=True, in_dtypes='float32'}[native:0]
FAIL tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py::test_QR_param_23_{shape=(0, 3), mode='reduced', skip_backward_test=True, skip_double_backward_test=True, in_dtypes='float32'}[native:0]
FAIL tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py::test_QR_param_20_{shape=(0, 3), mode='complete', skip_backward_test=True, skip_double_backward_test=True, in_dtypes='float32'}[native:0]
FAIL tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py::test_QR_param_29_{shape=(0, 3), mode='reduced', skip_backward_test=True, skip_double_backward_test=True, in_dtypes='float64'}[native:0]
FAIL tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py::test_QR_param_0_{shape=(0, 3), mode='r', skip_backward_test=True, skip_double_backward_test=True, in_dtypes='float32'}[native:0]
FAIL tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py::test_QR_param_15_{shape=(0, 3), mode='raw', skip_backward_test=True, skip_double_backward_test=True, in_dtypes='float64'}[native:0]
[31m[1m========================= 8 failed, 87839 passed, 83514 skipped, 41 xfailed in 1122.44 seconds =========================[0m
```

[the remark this answers is not in the document]
Comment (IvanYashchuk) [reply tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py]: Starting from NumPy 1.13 empty arrays are allowed for linalg routines.

**Comment by IvanYashchuk** [comment]
TODO:
With shape (N, 0) and mode='complete' cuSOLVER does not return the same result as LAPACK for q matrix (it should be the identity matrix). The solution is to handle this case explicitly separately.

File: tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (IvanYashchuk) [reply tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py]: Actually, it is not the case for all routines
Some, like `linalg.qr`, got this working only from 1.16
https://github.com/numpy/numpy/blob/master/doc/release/1.16.0-notes.rst#linalglstsq-linalgqr-and-linalgsvd-now-work-with-empty-arrays

[the remark this answers is not in the document]
Comment (IvanYashchuk) [reply tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py]: It is fixed now.

File: tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py [hunk 7]
```diff
@@ -9,6 +9,8 @@
 from chainerx_tests import dtype_utils
 from chainerx_tests import op_utils
 
+from distutils.version import LooseVersion
```
Comment (hvy) [inline tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py]: A minor suggestion (although I just noticed we're already using LooseVersion elsewhere) to use NumpyVersion instead to check the NumPy version as in https://github.com/chainer/chainer/blob/master/chainer/functions/array/get_item.py#L13.

File: chainerx_cc/chainerx/native/native_device/linalg.cc [hunk 8]
```diff
@@ -241,6 +279,85 @@ void InverseImpl(const Array& a, const Array& out) {
...
 
+template <typename T>
+void QrImpl(const Array& a, const Array& q, const Array& r, const Array& tau, QrMode mode) {
+    Device& device = a.device();
+    Dtype dtype = a.dtype();
+
+    int64_t m = a.shape()[0];
+    int64_t n = a.shape()[1];
+    int64_t k = std::min(m, n);
+    int64_t lda = std::max(static_cast<int64_t>(1), m);
```
Comment (hvy) [inline chainerx_cc/chainerx/native/native_device/linalg.cc]: Ditto.

File: chainerx_cc/chainerx/native/native_device/linalg.cc [hunk 9]
```diff
@@ -241,6 +279,85 @@ void InverseImpl(const Array& a, const Array& out) {
...
+
+    auto r_ptr = static_cast<T*>(internal::GetRawOffsetData(r_temp));
+    auto tau_ptr = static_cast<T*>(internal::GetRawOffsetData(tau));
+
+    int info;
+    int64_t buffersize_geqrf = -1;
+    T work_query_geqrf;
+    Geqrf(m, n, r_ptr, lda, tau_ptr, &work_query_geqrf, buffersize_geqrf, &info);
+    buffersize_geqrf = static_cast<int64_t>(work_query_geqrf);
+    buffersize_geqrf = std::max({static_cast<int64_t>(1), n, buffersize_geqrf});  // buffersize >= n >= 1
```
Comment (hvy) [inline chainerx_cc/chainerx/native/native_device/linalg.cc]: Ditto.

File: chainerx_cc/chainerx/cuda/cuda_device/linalg.cu [hunk 10]
```diff
@@ -243,6 +327,95 @@ void SolveImpl(const Array& a, const Array& b, const Array& out) {
...
+        mc = k;
+        q_shape = Shape{n, m};
+    }
+    Array q_temp = Empty(q_shape, dtype, device);
+
+    // cuSOLVER does not return correct result in this case
+    if (mode == QrMode::kComplete && n == 0) {
+        device.backend().CallKernel<IdentityKernel>(q);
+        return;
+    }
```
Comment (hvy) [inline chainerx_cc/chainerx/cuda/cuda_device/linalg.cu]: How about having this early return at the top?

File: tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py (hunk 7 above)
Comment (IvanYashchuk) [reply tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py]: I've changed that. Thank you!

File: chainerx_cc/chainerx/cuda/cuda_device/linalg.cu [hunk 11]
```diff
@@ -243,6 +327,95 @@ void SolveImpl(const Array& a, const Array& b, const Array& out) {
...
 
+template <typename T>
+void QrImpl(const Array& a, const Array& q, const Array& r, const Array& tau, QrMode mode) {
+    Device& device = a.device();
+    Dtype dtype = a.dtype();
+
+    int64_t m = a.shape()[0];
+    int64_t n = a.shape()[1];
+    int64_t k = std::min(m, n);
+    int64_t lda = std::max(static_cast<int64_t>(1), m);
```
[the remark this answers is not in the document]
Comment (IvanYashchuk) [reply chainerx_cc/chainerx/cuda/cuda_device/linalg.cu]: I don't know. Thank you for the suggestion!
Fixed.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
LGTM, thanks for the long running PR! Let's just check the CI.

**Comment by hvy** [comment]
Seems like we need to fix an edge case for empty inputs https://jenkins.preferred.jp/job/chainer/job/chainer_pr/1873/

**Comment by IvanYashchuk** [comment]
> Seems like we need to fix an edge case for empty inputs https://jenkins.preferred.jp/job/chainer/job/chainer_pr/1873/

I don't know what is causing this. The error is not reproduced on my computer with CUDA 10.1

From CUDA changelog:
>In earlier releases, cuBLAS GEMM calls might randomly crash when running multiple host threads sharing one cuBLAS handle despite adhering to recommended usage in Thread Safety. This bug affects cuBLAS in earlier CUDA releases, and is fixed in CUDA 10.1.
This bug affects cuSOLVER dense linear algebra functionality in CUDA 10.0, and is fixed in CUDA 10.1. An internal routine (LARFT) in cuSOLVER dense linear algebra has a race condition if the user stream is not null. Although this routine is not in the cuSOLVER public API, it affects all routines based on householder reflection, including ORGBR, ORMTR, ORGTR, ORMQR, ORMQL, ORGQR AND SYEVD, SYGVD.

ChainerX implementation of QR decomposition calls ORGQR, which is in the list of affected routines. That might or might not cause our troubles. Anyway, I've added an early return for the case of an empty input array.

**Comment by hvy** [comment]
Thanks for the fix, let me verify it. As for the release note citation above, I also read it but we aren't using non-null streams AFAIK.

**Comment by hvy** [comment]
Removing `test-and-merge` as merging is blocked until the release on Thu.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Still seeing failures on Jenkins. Could it be for non-"complete" cases?

**Comment by IvanYashchuk** [comment]
I suppose the reason for failure was `complete` mode && `m==0` which was not covered by an early return.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Thanks for the long running PR as always, LGTM!
